kernel/cls/cls_macros/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+    _init: Expr,
+}
+
+impl Parse for CpuLocal {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let attributes = input.call(Attribute::parse_outer)?;
+        let visibility: Visibility = input.parse()?;
+        input.parse::<Token![static]>()?;
+        let name: Ident = input.parse()?;
+        input.parse::<Token![:]>()?;
+        let ty: Type = input.parse()?;
+        input.parse::<Token![=]>()?;
+        let _init: Expr = input.parse()?;
+        input.parse::<Token![;]>()?;
+        Ok(CpuLocal {
+            attributes,
+            visibility,
+            name,
+            ty,
+            _init,
+        })
+    }
+}
+
+/// A macro for declaring CPU-local variables.
+///
+/// Variables must be an unsigned integer, bar `u128`.
+///
+/// The initialisation expression has no effect; to set the initial value,
+/// `per_cpu::PerCpuData::new` must be modified.
+#[proc_macro_attribute]
+pub fn cpu_local(args: TokenStream, input: TokenStream) -> TokenStream {
+    // if !args.is_empty() {
+    //     Diagnostic::spanned(
+    //         Span::call_site(),
+    //         Level::Error,
+    //         "malformed `cpu_local` attribute input",
+    //     )
+    //     .help("must be of the form `#[cpu_local]`")
+    //     .emit();
+    //     return TokenStream::new();
+    // }
+
+    let offset = if let Ok(i) = syn::parse::<LitInt>(args.clone()) {
+        i
+    } else {
+        let span = args
+            .into_iter()
+            .map(|tt| tt.span())
+            .reduce(|a, b| a.join(b).unwrap())
+            .unwrap_or_else(Span::call_site);
+        Diagnostic::spanned(span, Level::Error, "invalid offset").emit();
+        return TokenStream::new();
+    };
+
+    let CpuLocal {
+        attributes,
+        visibility,
+        name,
+        ty,
+        _init,
+    } = parse_macro_input!(input as CpuLocal);
+
+    let attributes = attributes.iter().map(|attribute| quote! { #attribute });
+
+    let struct_name = Ident::new(
+        &format!("CPU_LOCAL_{name}").to_case(Case::Pascal),
+        Span::call_site().into(),
+    );
+
+    let ((x64_asm_width, x64_reg_class), (aarch64_reg_modifier, aarch64_instr_width)) =
+        match ty.to_token_stream().to_string().as_ref() {
+            "u8" => (("byte", quote! { reg_byte }), (":w", "b")),
+            "u16" => (("word", quote! { reg }), (":w", "w")),
+            "u32" => (("dword", quote! { reg }), (":w", "")),
+            "u64" => (("qword", quote! { reg }), ("", "")),
+            _ => {
+                Diagnostic::spanned(ty.span().unwrap(), Level::Error, "invalid type")
+                    .help("CPU locals only support these types: `u8`, `u16`, `u32`, `u64`")
+                    .emit();
+                return TokenStream::new();
+            }
+        };
+
+    let x64_width_modifier = format!("{x64_asm_width} ptr ");
+    let x64_cls_location = format!("gs:[{offset}]");
+
+    quote! {
+        #(#attributes)*
+        #[thread_local]
```

Comment:
No. This is the list of symbols in the `preemption` elf file
```
Symbol table '.symtab' contains 24 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 0000000000000000     0 SECTION LOCAL  DEFAULT    2 .text
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT    4 .rodata
     3: 0000000000000000     0 SECTION LOCAL  DEFAULT    8 .gcc_except_table
     4: 0000000000000000    12 FUNC    LOCAL  DEFAULT    2 <cpu::CpuId as core::fmt::Display>::fmt
     5: 0000000000000010    12 FUNC    LOCAL  DEFAULT    2 core::ptr::drop_in_place<preemption::PreemptionGuard>
     6: 0000000000000020    24 FUNC    LOCAL  DEFAULT    2 core::ptr::drop_in_place<sync::rw_lock::RwLockWriteGuard<apic::LocalApic,sync_irq::DisableIrq>>
     7: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND cpu::arch::current_cpu
     8: 0000000000000240   219 FUNC    GLOBAL DEFAULT    2 preemption::hold_preemption_no_timer_disable
     9: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND core::panicking::panic_cannot_unwind
    10: 0000000000000560    14 FUNC    GLOBAL DEFAULT    2 preemption::preemption_enabled
    11: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND apic::LocalApic::enable_lvt_timer
    12: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND apic::get_my_apic
    13: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND _Unwind_Resume
    14: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND core::panicking::panic_fmt
    15: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND <irq_safety::held_interrupts::HeldInterrupts as core::ops::drop::Drop>::drop
    16: 0000000000000330     3 FUNC    GLOBAL DEFAULT    2 preemption::PreemptionGuard::cpu_id
    17: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND core::option::expect_failed
    18: 0000000000000320     5 FUNC    GLOBAL DEFAULT    2 preemption::PreemptionGuard::preemption_was_enabled
    19: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND core::fmt::num::imp::<impl core::fmt::Display for u32>::fmt
    20: 0000000000000040   511 FUNC    GLOBAL DEFAULT    2 preemption::hold_preemption
    21: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND irq_safety::held_interrupts::hold_interrupts
    22: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND rust_eh_personality
    23: 0000000000000340   543 FUNC    GLOBAL DEFAULT    2 <preemption::PreemptionGuard as core::ops::drop::Drop>::drop
```

Reply:
Awesome, thanks for confirming. I would presume the symbol is completely removed since it's a ZST, and only appears by means of its functions that get inlined into the callsites